headers/backend/api/functions.h

```diff
@@ -0,0 +1,154 @@
+#ifndef FUNCTIONS_H
+#define FUNCTIONS_H
+
+#include "frontend/ast.h"
+#include "backend/values/varenv.h"
+
+/// SECTION: Macros
+
+#define FUNC_ARGV_MIN_SZ 4
+#define FUNC_ARGV_MAX_SZ 32
+
+/// SECTION: Type Decls.
+
+typedef enum en_function_type
+{
+    FUNC_NATIVE,  // wrapper for C function
+    FUNC_NORMAL,  // wrapper for AST nodes to traverse
+    FUNC_UNKNOWN  // reserved for bytecode usage??
+} FuncType;
+
+/// SECTION: Func Args Impl.
+
+typedef struct st_func_args
+{
+    unsigned short argc;
+    VarValue **args;
+} FuncArgs;
+
+typedef VarValue *(*NativeFunc)(struct st_func_args *args);
+
+FuncArgs *funcargs_create(unsigned short argc);
+
+/**
+ * @brief Unbinds referencing ptrs to arg values.
+ * @param argv
+ */
+void funcargs_dispose(FuncArgs *argv);
+
+/**
+ * @brief Frees up memory of stored referencing ptrs to arg values.
+ * @param argv 
+ * @note Use only when there is no interpreter scope (native func calls!)
+ */
+void funcargs_destroy(FuncArgs *argv);
+
+int funcargs_set_at(FuncArgs *argv, unsigned short index, VarValue *arg);
+
+VarValue *funcargs_get_at(const FuncArgs *argv, unsigned short index);
+
+/// SECTION: Func Params Impl.
+
+typedef struct st_func_params
```

Comment:
Currently, I have duplicate code for dynamic arrays within some structures: the `FuncParams` array, the `Script` structure, and AST nodes such as those for compound statements (blocks). Thus, I should add a to-do for refactoring that duplicate code by placing it into a `void*` dynamic array implementation. A future refactor will help keep the code cleaner and easier to extend.